### ISSUE TITLE
Losing 1 day when saving member expiration/enddate

### ISSUE THREAD

**Issue by eighty20results** [strangerstudios/paid-memberships-pro#907]
As has been reported a bunch of times in the PMPro Member Support forum, due to the use of `UNIX_TIMESTAMP()` when saving date values, the timezone settings of the WordPress environment will result in a member _losing_ a day every time their WP User profile is saved. 

I currently have a customer who's looking for a fix for this issue ASAP.  They're wanting to use the PMPro start/end date info as the foundation for a physical access solution which means that the failure to save the start/end date info per the timezone the WordPress admin can control - Timezone setting in WordPress - is causing real problems for them.

The problems are significant enough to consider replacing PMPro as their member management solution.

This issue is actually resolved in (pull request #581) with a combination of removing the `UNIX_TIMESTAMP()` SQL _and_ ensuring the data is returned as a 'seconds since epoch' format by using `strtotime()` on the returned value(s). The fixes were tested in my own test and production environments (across several sites). 

The fixes from that pull request that are relevant are the following changes:

https://github.com/strangerstudios/paid-memberships-pro/pull/581/commits/6ebf02145ae453b6c2d980ce1158ccf5da6a960f
https://github.com/strangerstudios/paid-memberships-pro/pull/581/commits/83494176e1ce3e317b04c0cb3844ff635ced23cd
https://github.com/strangerstudios/paid-memberships-pro/pull/581/commits/6b23124c9f253d63a5af1b86a011fb1de8507eff
https://github.com/strangerstudios/paid-memberships-pro/pull/581/commits/becdf3694393f029fbf1cca8a94cae97b63d5469
https://github.com/strangerstudios/paid-memberships-pro/pull/581/commits/f8b41d64ea3127930813b89fe8fe97f722c6e975
https://github.com/strangerstudios/paid-memberships-pro/pull/581/commits/f36ff818ed7c906e68ed16ba8b87f6bcc8215ee4

**Comment by ideadude** [comment]
Thanks staying on top of this and pulling out the specific commits. I'll set aside time next week to get through these.

**Comment by talkytoaster** [comment]
I'm seeing the same issue, anytime I edit a member they lose a day, this is rather annoying and should be simple to fix? Please put the fix in to production...

**Comment by arnalyse** [comment]
We have the same problem with one of our clients. We came up with a similar solution to @eighty20results. We got rid of the `UNIX_TIMESTAMP` and let `DateTime()` calculate the correct timezone when reading the result from the database: 

```diff
--- a/docroot/wp-content/plugins/paid-memberships-pro/classes/class.memberorder.php
+++ b/docroot/wp-content/plugins/paid-memberships-pro/classes/class.memberorder.php
@@ -78,7 +78,7 @@
 				return false;
 
 			$gmt_offset = get_option('gmt_offset');
-			$dbobj = $wpdb->get_row("SELECT *, UNIX_TIMESTAMP(timestamp) + " . ($gmt_offset * 3600) . "  as timestamp FROM $wpdb->pmpro_membership_orders WHERE id = '$id' LIMIT 1");
+			$dbobj = $wpdb->get_row("SELECT * FROM $wpdb->pmpro_membership_orders WHERE id = '$id' LIMIT 1");
 
 			if($dbobj)
 			{
@@ -135,7 +135,9 @@
 				$this->gateway_environment = $dbobj->gateway_environment;
 				$this->payment_transaction_id = $dbobj->payment_transaction_id;
 				$this->subscription_transaction_id = $dbobj->subscription_transaction_id;
-				$this->timestamp = $dbobj->timestamp;
+
+				$this->timestamp = DateTime::createFromFormat('Y-m-d H:i:s', $dbobj->timestamp)->getTimestamp();
+
 				$this->affiliate_id = $dbobj->affiliate_id;
 				$this->affiliate_subid = $dbobj->affiliate_subid;
```

Is there any way we can help to get the fixes into the next release? :)

**Comment by mossifer** [comment]
VERY annoying bug - our members are losing days in their memberships! HELP

**Comment by eighty20results** [comment]
@mossifer I can understand the frustration. There _is_ a sub-par workaround (add +1 to the day on the date when editing/updating the user's Profile page as an admin)...

Obviously, I don't know how scalable that is for your site, so YMMW.

**Comment by ideadude** [comment]
Fixed in 2.3 release. Please open a new issue if you notice any errors.
